R/fcast.R

```diff
@@ -211,7 +210,12 @@ dcast.data.table = function(data, formula, fun.aggregate = NULL, sep = "_", ...,
     }
     maplen = vapply_1i(mapunique, length)
     idx = do.call("CJ", mapunique)[map, 'I' := .I][["I"]] # TO DO: move this to C and avoid materialising the Cross Join.
-    ans = .Call(Cfcast, lhs, val, maplen[[1L]], maplen[[2L]], idx, fill, fill.default, is.null(fun.call))
+    fill.default = NULL
+    some_fill = anyNA(idx)
+    if (run_agg_funs && is.null(fill) && some_fill) {
+      fill.default = dat_for_default_fill[0L][, maybe_err(eval(fun.call))]
+    }
+    ans = .Call(Cfcast, lhs, val, maplen[[1L]], maplen[[2L]], idx, fill, fill.default, is.null(fun.call), some_fill)
```

Comment:
Probably something to examine as a follow-up but I'm surprised we need all of `fill`, `fill.default`, and `some_fill` as arguments, it seems we should be able to remove at least one. Unless you're sure all three are necessary, I'll write the issue.

Reply:
yes I think they are all necessary, I was surprised too. 
in particular I tried replacing if(some_fill) in C code on !isNull(fill_d) but that did not work (there are times when fill_d is NULL but there is apparently something to fill still, not exactly sure why)